### PR TITLE
Merging API cleanup branch

### PR DESCRIPTION
**Pull request by benvanik**
This cleans up most of the compiler warnings we have and adds string parsing/formatting to the C API. This will allow us to replace the existing legacy types like iree/base/shaped_buffer.h and iree/base/shape.h and the glue in vm_util.h with the real C versions. The changes here result in a smaller compiled codesize (less giant template type expansion) and will make things really easy to use from java/etc. Most of the existing shaped buffer tests were ported over though with the new code organization far less duplication was required (and someone wasn't trying to go for readability at the time, which helps).

There's also some iteration on the C++ binding (that relies only on the C API). More to come on that.

### CODE REVIEW

File: iree/hal/api_string_util_test.cc [hunk 1]
```diff
@@ -0,0 +1,1062 @@
...
+
+TEST(ShapeStringUtilTest, ParseShape) {
+  EXPECT_THAT(ParseShape(""), IsOkAndHolds(Eq(Shape{})));
+  EXPECT_THAT(ParseShape("0"), IsOkAndHolds(Eq(Shape{0})));
+  EXPECT_THAT(ParseShape("1"), IsOkAndHolds(Eq(Shape{1})));
+  EXPECT_THAT(ParseShape("1x2"), IsOkAndHolds(Eq(Shape{1, 2})));
+  EXPECT_THAT(ParseShape(" 1 x 2 "), IsOkAndHolds(Eq(Shape{1, 2})));
+  EXPECT_THAT(ParseShape("1x2x3x4x5"), IsOkAndHolds(Eq(Shape{1, 2, 3, 4, 5})));
+  EXPECT_THAT(ParseShape("1x2x3x4x5x6x7x8x9"),
+              IsOkAndHolds(Eq(Shape{1, 2, 3, 4, 5, 6, 7, 8, 9})));
```
Comment (ScottTodd) [inline iree/hal/api_string_util_test.cc:550]: Would be good to run fuzz tests on these APIs as well.

File: iree/hal/api_string_util_test.cc [hunk 2]
```diff
@@ -0,0 +1,1062 @@
...
+namespace iree {
+namespace hal {
+namespace {
+
+using ::iree::testing::status::IsOkAndHolds;
+using ::iree::testing::status::StatusIs;
+using ::testing::ElementsAre;
+using ::testing::Eq;
+
+// TODO(benvanik): move these utils to C++ bindings.
```
Comment (ScottTodd) [inline iree/hal/api_string_util_test.cc:33]: :P sure are a lot of utils in this `test.cc` file
thankfully they are pretty thin wrappers, but we want to ensure that the test is exercising the underlying C API / impl faithfully

File: iree/hal/api_string_util_test.cc [hunk 3]
```diff
@@ -0,0 +1,1062 @@
...
+  EXPECT_THAT(ParseElement<int32_t>("asdfasdf"),
+              StatusIs(StatusCode::kInvalidArgument));
+  EXPECT_THAT(ParseElement<uint32_t>("asdfasdf"),
+              StatusIs(StatusCode::kInvalidArgument));
+  EXPECT_THAT(ParseElement<float>("asdfasdf"),
+              StatusIs(StatusCode::kInvalidArgument));
+  EXPECT_THAT(ParseElement<double>("asdfasdf"),
+              StatusIs(StatusCode::kInvalidArgument));
+
+  EXPECT_THAT(ParseElement<int8_t>("🌮"),
```
Comment (ScottTodd) [inline iree/hal/api_string_util_test.cc:699]: 🌮

File: iree/hal/api.cc [hunk 4]
```diff
@@ -606,9 +1097,19 @@ IREE_API_EXPORT iree_host_size_t IREE_API_CALL iree_hal_buffer_view_shape_dim(
   return buffer_view->shape[index];
 }
 
+IREE_API_EXPORT iree_host_size_t
+iree_hal_buffer_view_element_count(const iree_hal_buffer_view_t* buffer_view) {
+  CHECK(buffer_view) << "NULL buffer_view handle";
```
Comment (ScottTodd) [inline iree/hal/api.cc:1102]: return `IREE_STATUS_INVALID_ARGUMENT` instead?

**Review by ScottTodd — APPROVED** [review]
nit: "Fixing test file name convention mismatch." commit could be in a separate PR.

File: iree/hal/api_string_util_test.cc (hunk 2 above)
Comment (benvanik) [reply iree/hal/api_string_util_test.cc:33]: yep, this is a testing ground for them - see #1846

File: iree/hal/api_string_util_test.cc [hunk 5]
```diff
@@ -0,0 +1,1062 @@
...
+StatusOr<Shape> ParseShape(absl::string_view value) {
+  Shape shape(6);
+  iree_host_size_t actual_rank = 0;
+  iree_status_t status;
+  do {
+    status =
+        iree_hal_parse_shape(iree_string_view_t{value.data(), value.size()},
+                             shape.size(), shape.data(), &actual_rank);
+    shape.resize(actual_rank);
+  } while (iree_status_is_out_of_range(status));
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/hal/api_string_util_test.cc:47]: saves on code-size and risk of typos

File: iree/hal/api.cc (hunk 4 above)
Comment (benvanik) [reply iree/hal/api.cc:1102]: this is an accessor property and doesn't return a status (it's equivalent to a pointer dereference)

**Pull request closed without merging**